### PR TITLE
READY : Chat `push_msg()` implementation for MultiplayerGame.

### PR DESCRIPTION
**Pull request by al-mehedi**
Currently broadcasts to all clients. Need suggestions on how to stream chat to the appropriate client.

### CODE REVIEW

File: module/server/src/rpc_server.rs [hunk 1]
```diff
@@ -201,7 +201,34 @@ impl Chess for ChessRpcServer
...
+    let player = msg_req.player;
+
+    if let Some(player) = player {
+      let msg = MultiplayerMessage::new(player.player_id, msg_req.text);
+
+      let streams = self.streams.lock().expect("Failed to lock the streams mutex");
+
+      // Broadcasts chat message to all clients.
+      // TODO; send message to appropriate recipient
+      // How to identify opponent remote address???
```
Comment (dmvict) [inline module/server/src/rpc_server.rs:216]: Good implementation. 

The chat channels should be coupled with the instance of Game. So, if some user sends message, then message should contain the ID of the game.

Comment (al-mehedi) [reply module/server/src/rpc_server.rs:216]: https://github.com/Wandalen/game_chess/blob/e0d028f10e0bb32c74459863b789116ca45a6e83/module/multiplayer/proto/chess_game.proto#L26
Can you please explain!

Currently, there's no game instance attached with the multiplayer game, Should I add a field to MultiplayerGame for game instance! and, should our chat messages will be real-time or be saved within the game instance so the players will be pulling it when appropriate?

Comment (dmvict) [reply module/server/src/rpc_server.rs:216]: Yes, you can attach instance of Chat to the Game. It's main idea to have game chat in format:
```
Player1 |
        | => Chat
Player2 |
```
If feature `Spectators` is implemented
```
Player1    |
Player2    | => Chat
Spectators |
```
Should game be a real-time or save history? Feel free and implement chat you like.

Comment (al-mehedi) [reply module/server/src/rpc_server.rs:216]: Should I disregard the chat messages for streams for the time being?

Comment (dmvict) [reply module/server/src/rpc_server.rs:216]: Please, explain the idea about these streams, and I will answer your question.

Comment (Wandalen) [reply module/server/src/rpc_server.rs:216]: Let's avoid coupling of basic playing functionality( GameUpdate ) and chat feature, please.

Comment (dmvict) [reply module/server/src/rpc_server.rs:221]: I've got more details and corrected description of task #60 

Please, take it into account.

File: module/cui/src/main.rs [hunk 2]
```diff
@@ -459,3 +462,41 @@ pub async fn command_online_game_join()
...
   }
 }
+
+///
+/// Command to send message to opponent.
+///
+pub async fn command_online_game_send_msg()
+{
+  if let Ok(mut chess_client) = chess_client::ChessClient::connect("http://127.0.0.1:1313").await {
+    let player_id = wca::input::ask("Input Player ID");
```
Comment (dmvict) [inline module/cui/src/main.rs:472]: The PR looks like is ready to be accepted. 

Can you decrease number of questions for the player?

Comment (dmvict) [reply module/cui/src/main.rs:472]: I understand it, but the client or instance of game should know Player ID. 

Let's imagine that we used this interface. 

```
.online.msg 
my_id
destination_game_id
my msg
```

I think user should no input nor Player ID not Game ID.
The interface:

```
.online.msg 
my msg
```

Spectators can have another interface.

 Can you implement it?

Comment (al-mehedi) [reply module/cui/src/main.rs:472]: Yes, I believe I could implement it. Let me try

Comment (dmvict) [reply module/cui/src/main.rs:472]: Thank you :) 

I believe that user experience matter :)

File: module/cui/src/main.rs [hunk 3]
```diff
@@ -78,6 +78,36 @@ use game_chess_core::*;
 #[allow(unused_imports)]
 use game_chess_client::*;
 
+///
+/// Simple Session for Online MultiplayerGame
```
[the remark this answers is not in the document]
Comment (dmvict) [reply module/cui/src/main.rs:82]: I've tested the app with messaging. 

Really cool way to communicate using the CUI. 

I think one improvement the messaging required. 

Now, when user typed 
```
.online.msg.read
```
output prints only opponents  messages. 

I think that user should see all messages history that is messages of Player1 and Player2. 

Can you add this behavior?

[the remark this answers is not in the document]
Comment (al-mehedi) [reply module/cui/src/main.rs:82]: Yes, I filtered out messages that are sent from the same client, let me modify this behavior.

[the remark this answers is not in the document]
Comment (al-mehedi) [reply module/cui/src/main.rs:82]: Now `.online.msg.read` prints out chat history for both players.

[the remark this answers is not in the document]
Comment (dmvict) [reply module/cui/src/main.rs:82]: Ok, add label `READY : ` at start of the PR name and @Wandalen should accept it.

**Comment by al-mehedi** [comment]
@dmvict Please let me know if you have any suggestions on the recent commit. give it a try :)

**Pull request merged**